### PR TITLE
Update Golang to 1.12.1

### PR DESCRIPTION
**Pull request by kvch**
Changes in 1.12: https://golang.org/doc/go1.12
Changes in minor release: https://github.com/golang/go/issues?q=milestone%3AGo1.12.1

Updated vendored packages:
- `github.com/tsg/gopacket/pcap`
- `github.com/docker/docker/pkg/*`
- `github.com/docker/go-connections`

New vendored packages (as required by dependencies):
- `github.com/konsorten/go-windows-terminal-sequences`

Removed vendored package:
- `github.com/Sirupsen/logrus`: `github.com/sirupsen/logrus` is used and it leads to a case insensitive collision, required by docker dependencies

TODO
- [x] use documented Cgo types in `github.com/tsg/gopacket/pcap` [Cgo changes](https://golang.org/doc/go1.12#cgo)
- [x] fix recovering from panic in `logp`

Possibly interesting PR to devs @elastic/apm-server

### CODE REVIEW

**Comment by ruflin** [comment]
Went through the changelog and also stumbled over TLS changes: https://golang.org/doc/go1.12 @ph I don't think it will have an affect but you should double check.

Overall +1 on going to 1.12.

**Comment by kvch** [comment]
@ruflin I would like to add support for TLS 1.3

File: metricbeat/Dockerfile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-FROM golang:1.10.3
```
Comment (kvch) [inline metricbeat/Dockerfile]: @ruflin Do you know why the version of Golang differed from other Beats versions? Is it ok to update it?
I haven't seen any related failures yet.

**Comment by ph** [comment]
We might want to look at using this:

```
The compiler now accepts a -lang flag to set the Go language version to use. For example, -lang=go1.8 causes the compiler to emit an error if the program uses type aliases, which were added in Go 1.9. Language changes made before Go 1.12 are not consistently enforced. 
```

**Comment by ph** [comment]
> On Linux, the runtime now uses MADV_FREE to release unused memory. This is more efficient but may result in higher reported RSS. The kernel will reclaim the unused data when it is needed. To revert to the Go 1.11 behavior (MADV_DONTNEED), set the environment variable GODEBUG=madvdontneed=1. 

Something to keep in mind on SDH

File: libbeat/logp/core_test.go [hunk 2]
```diff
@@ -167,6 +167,8 @@ func TestLoggerLevel(t *testing.T) {
 }
 
 func TestRecover(t *testing.T) {
+	t.Skip("go.uber.org/zap does not support Golang 1.12. See: https://github.com/uber-go/zap/issues/682")
+
```
Comment (ph) [inline libbeat/logp/core_test.go]: @kvch What's up with this? I am trying to understand the impact of the above issue on us if we are upgrading. My questions are:

- Did you get the same errors as the issue above?
- I currently don't see any tracking on the zap repo.
- Looking at the golang issue some commits where done to address the problem.

This look like an issue concerning the extracted line numbers when a panic occurs.

**Comment by ph** [comment]
Looking at the TLSv1.3, from my understanding you need to opt-in via an environment variables, not sure if can force it via a TLSConfig.

**Comment by graphaelli** [comment]
While it hasn't been kept up so far, should the various golang.org/x packages be kept in sync with this update too?  These are currently vendored:

```
golang.org/x/net/bpf@release-branch.go1.9
golang.org/x/net/context@release-branch.go1.9
golang.org/x/net/context/ctxhttp@release-branch.go1.9
golang.org/x/net/http2@release-branch.go1.9
golang.org/x/net/http2/hpack@release-branch.go1.9
golang.org/x/net/icmp@release-branch.go1.9
golang.org/x/net/internal/iana@release-branch.go1.9
golang.org/x/net/ipv4@release-branch.go1.9
golang.org/x/net/ipv6@release-branch.go1.9
golang.org/x/net/lex/httplex@release-branch.go1.9
golang.org/x/net/proxy@release-branch.go1.9
golang.org/x/net/publicsuffix@release-branch.go1.9
golang.org/x/tools/cmd/goimports@release-branch.go1.10
golang.org/x/tools/go/ast/astutil@release-branch.go1.10
golang.org/x/tools/go/buildutil@release-branch.go1.9
golang.org/x/tools/go/loader@release-branch.go1.9
golang.org/x/tools/imports@release-branch.go1.10
golang.org/x/tools/internal/fastwalk@release-branch.go1.10
golang.org/x/tools/internal/gopathwalk@release-branch.go1.10
```

**Comment by ph** [comment]
@graphaelli correct, I think we were just updating them as needed or when a bug was detected (mostly following what ain't broke don't fix it..). I think we should clarify that strategy and I would be +1 to keep them in sync. Wdyt @ruflin?

File: libbeat/logp/core_test.go (hunk 2 above)
Comment (kvch) [reply libbeat/logp/core_test.go]: The caller of the logging function is not returned correctly. Their unit tests also catch this problem, as shown in the issue I have linked.

This is the error I am getting when running the test:
```
--- FAIL: TestRecover (0.00s)
    core_test.go:180:
                Error Trace:    core_test.go:180
                                                        core_test.go:188
                Error:          Not equal:
                                expected: "logp/core_test.go"
                                actual  : "testing/testing.go"

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1 +1 @@
                                -logp/core_test.go
                                +testing/testing.go
                Test:           TestRecover
FAIL
exit status 1
FAIL    github.com/elastic/beats/libbeat/logp   0.003s
```
The caller is incorrect.

I have tried first to see if updating the package helps. Unfortunately, it does not.

I have doubts whether it is going to be fixed soon. I would fork a repo and fix it. Until the fix is added to upstream, I would use Beats with the fork. WDYT?
I think Golang 1.12 brings a lot of value which is worth the additional effort.

**Comment by kvch** [comment]
@ph We could document that adding the environment variable lets you use TLSv3. But it's not so user friendly. Or we could add the feature regardless and release it officially when we update Go to 1.13. :D

@graphaelli @ph :+1: for keeping these packages in sync

File: libbeat/logp/core_test.go (hunk 2 above)
Comment (kvch) [reply libbeat/logp/core_test.go]: It's not a bug. I had to do a minor follow-up in `logp/global.go` to decrease the number of caller skipped, as the panic tracebacks had been changed. (https://go-review.googlesource.com/c/go/+/153477/)

**Comment by ruflin** [comment]
+1 to keep the libs in sync.

File: vendor/github.com/tsg/gopacket/pcap/pcap.go [hunk 3]
```diff
@@ -170,7 +170,7 @@ type InterfaceAddress struct {
 // BPF is a compiled filter program, useful for offline packet matching.
 type BPF struct {
 	orig string
-	bpf  _Ctype_struct_bpf_program // takes a finalizer, not overriden by outsiders
+	bpf  C.struct_bpf_program // takes a finalizer, not overriden by outsiders
```
Comment (ruflin) [inline vendor/github.com/tsg/gopacket/pcap/pcap.go]: Not really familiar with this change. @tsg could you have a look?

File: vendor/github.com/tsg/gopacket/pcap/pcap_tester.go [hunk 4]
```diff
@@ -1,107 +0,0 @@
...
-//
-// Use of this source code is governed by a BSD-style license
-// that can be found in the LICENSE file in the root of the source
-// tree.
-
-// +build ignore
-
-// This binary tests that PCAP packet capture is working correctly by issuing
-// HTTP requests, then making sure we actually capture data off the wire.
-package main
```
Comment (ruflin) [inline vendor/github.com/tsg/gopacket/pcap/pcap_tester.go]: Is the removal of this file related to 1.12.1?

File: vendor/github.com/tsg/gopacket/pcap/pcap.go (hunk 3 above)
Comment (kvch) [reply vendor/github.com/tsg/gopacket/pcap/pcap.go]: He merged the PR in his own repo. So he already accepted it. :)

File: vendor/github.com/tsg/gopacket/pcap/pcap_tester.go (hunk 4 above)
Comment (kvch) [reply vendor/github.com/tsg/gopacket/pcap/pcap_tester.go]: It was removed by govendor when I ran `fetch`. I thought it was removed from the original repo... But now I see that it's still there. Let me look into why the tool get rid of this file.

File: metricbeat/Dockerfile (hunk 1 above)
Comment (ruflin) [reply metricbeat/Dockerfile]: This is odd, all Beats should be in sync :-( I assume this had only an affect on testing and not packaging, meaning in 7.0 all Beats are packaged with the same Golang version.

Please update, I assume we missed it in the last Golang update.

Comment (kvch) [reply metricbeat/Dockerfile]: It was downgraded here: https://github.com/elastic/beats/pull/9242/files#diff-69f23758930df5eda4eeb23437b8a2ccR1

It seems like an accidental change. Or at least I haven't seen any reasoning why it was done.

File: vendor/github.com/tsg/gopacket/pcap/pcap_tester.go (hunk 4 above)
Comment (ruflin) [reply vendor/github.com/tsg/gopacket/pcap/pcap_tester.go]: The reason could be that its `package main`. My first gest was that it's a test file, but it's a script.

File: metricbeat/Dockerfile (hunk 1 above)
Comment (ruflin) [reply metricbeat/Dockerfile]: @kvch Thanks for investigation. @sayden I left a comment on the old PR, I wonder if any of these changes should have happened there.

**Comment by ruflin** [comment]
CI Failure looks related.

File: metricbeat/Dockerfile (hunk 1 above)
Comment (andrewkroh) [reply metricbeat/Dockerfile]: It probably was a merge gone bad.

Comment (ruflin) [reply metricbeat/Dockerfile]: @sayden Can you please open a PR with reverting to the old Dockerfile?

File: CHANGELOG.next.asciidoc [hunk 5]
```diff
@@ -10,6 +10,34 @@ https://github.com/elastic/beats/compare/v7.0.0-alpha2...master[Check the HEAD d
...
+- ILM will be available by default if Elasticsearch > 7.0 is used. {pull}10347[10347]
+- Allow Central Management to send events back to kibana. {issue}9382[9382]
+- Initialize the Paths before the keystore and save the keystore into `data/{beatname}.keystore`. {pull}10706[10706]
+- Add `cleanup_timeout` option to docker autodiscover, to wait some time before removing configurations after a container is stopped. {issue]10374[10374] {pull}10905[10905]
+- On Google Cloud Engine (GCE) the add_cloud_metadata will now trim the project
+  info from the cloud.machine.type and cloud.availability_zone. {issue}10968[10968]
+- Empty `meta.json` file will be treated as a missing meta file. {issue}8558[8558]
+- Rename `migration.enabled` config to `migration.6_to_7.enabled`. {pull}11284[11284]
+- Beats Xpack now checks for Basic license on connect. {pull}11296[11296]
+- Update to Golang 1.12.1. {pull}11330[11330]
```
Comment (urso) [inline CHANGELOG.next.asciidoc]: Looks like you need to rebase and fix the changelog?

Comment (kvch) [reply CHANGELOG.next.asciidoc]: Updated. Thanks.

**Comment by jordansissel** [comment]
Can this be backported into backport into 7.0/7.x? I only see it in master. (Context: I ran into this tonight)

**Pull request merged**